### DISCUSSION TITLE
getting this error torch.cuda.OutOfMemoryError: CUDA out of memory while running diffusers/examples/dreambooth/train_dreambooth.py

### DISCUSSION THREAD

**Discussion by akashlinux10may** [huggingface/diffusers#6932] [Q&A]
trying following command and getting the error

!accelerate launch diffusers/examples/dreambooth/train_dreambooth.py \
  --pretrained_model_name_or_path=CompVis/stable-diffusion-v1-4  \
  --instance_data_dir=diffusers/Dog \
  --output_dir=diffusers/Dog_SD \
  --instance_prompt="a photo of dog
![Screenshot 2024-02-11 101517](https://github.com/huggingface/diffusers/assets/66991234/840d8a73-3ecb-4035-a6f2-42342e94c820)
" \
  --resolution=512 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=1 \
  --learning_rate=5e-6 \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=400 \
  --push_to_hub

return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 114.00 MiB (GPU 0; 9.50 GiB total capacity; 1.13 GiB already allocated; 24.94 MiB free; 1.17 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF

**Comment by asomoza** [comment]
can you try adding the `--gradient_checkpointing` argument and test it?

**Reply by akashlinux10may** [reply]
Yes, I tried but still it's same. Can you give exact command which resolves this error?

**Reply by asomoza** [reply]
I can try in a collab, how much VRAM does your GPU have? I'm guessing 10GB but just to be sure.

Also here you can see the different arguments you'll need depending on your VRAM:

https://github.com/huggingface/diffusers/tree/main/examples/dreambooth

**Reply by akashlinux10may** [reply]
I am using on 128GB GPU.

**Reply by asomoza** [reply]
128GB? are you sure? can you tell me the GPU model?

**Reply by asomoza** [reply]
anyway this should work:

```sh
!accelerate launch diffusers/examples/dreambooth/train_dreambooth.py
  --mixed_precision="fp16" \
  --pretrained_model_name_or_path="CompVis/stable-diffusion-v1-4" \
  --instance_data_dir="diffusers/Dog" \
  --output_dir="diffusers/Dog_SD" \
  --instance_prompt="a photo of dog" \
  --resolution=512 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=1 \
  --gradient_checkpointing \
  --learning_rate=5e-6 \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=400 \
  --push_to_hub 
```

**Reply by akashlinux10may** [reply]
tried but again the above same error, and see my GPU env in the screenshot
![Uploading Screenshot 2024-02-11 125304.png…]()

**Reply by asomoza** [reply]
yeah, it seems a 10 or 12GB GPU, I'm making the colab to test it but I'm having trouble with a another issue, just if you want to try, add this too:

`--mixed_precision="fp16"`

with 10GB or 12GB you can't train in full precision.

**Reply by asomoza** [reply]
alright, I tested it in colab with this:

```sh
!accelerate launch train_dreambooth.py \
  --pretrained_model_name_or_path="CompVis/stable-diffusion-v1-4" \
  --instance_data_dir="/content/dog" \
  --instance_prompt="a photo of sks dog" \
  --resolution=512 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=1 \
  --gradient_checkpointing \
  --learning_rate=5e-6 \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=400 \
  --use_8bit_adam \
  --mixed_precision="fp16"
```

with this it uses:

![image](https://github.com/huggingface/diffusers/assets/5442875/28f8a7a1-7f48-42b9-a218-c6f7c822412c)

so if you have a 10GB card you can't train a full dreambooth model, you'll need to train a LoRA, it works with 16 GB since that's what a free colab gives.

**Reply by asomoza** [reply]
also in the documentation says you can train with 8GB using [deepspeed](https://www.deepspeed.ai/) but I have no experience with that so I can't help you.

**Reply by asomoza** [reply]
I got it to 9GB with  `--set_grads_to_none`

![image](https://github.com/huggingface/diffusers/assets/5442875/8eb37c5c-e6b7-40b2-a2d7-3319caf71657)

```sh
!accelerate launch train_dreambooth.py \
  --pretrained_model_name_or_path="CompVis/stable-diffusion-v1-4" \
  --instance_data_dir="/content/dog" \
  --instance_prompt="a photo of sks dog" \
  --resolution=512 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=1 \
  --gradient_checkpointing \
  --learning_rate=5e-6 \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=400 \
  --use_8bit_adam \
  --set_grads_to_none \
  --mixed_precision="fp16"
```

**Reply by akashlinux10may** [reply]
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 114.00 MiB (GPU 0; 9.50 GiB total capacity; 1.13 GiB already allocated; 24.94 MiB free; 1.17 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF

**Reply by akashlinux10may** [reply]
i tried in colab, there got this , RuntimeError: Expected is_sm80 || is_sm90 to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)
Steps:   0% 0/400 [00:03<?, ?it/s]

**Reply by asomoza** [reply]
yeah, that was the issue I was having, they changed the default torch to 2.1.0 which gives that error, I had to manually install the 2.2.0 version, you can use my colab as a guide if you want:

https://colab.research.google.com/drive/1gbRUa2FC59IkQTZDZuu4Ff8QxVH7gxaI?usp=sharing

**Reply by akashlinux10may** [reply]
while running your colab I am getting this error, ImportError: This example requires a source install from HuggingFace diffusers (see `[https://huggingface.co/docs/diffusers/installation#install-from-source`](https://huggingface.co/docs/diffusers/installation#install-from-source%60)), but the version found is 0.26.2.

**Reply by asomoza** [reply]
probably I did it before the release of that version, just replace this:

`!pip -q install diffusers`

with this:

`!pip install git+https://github.com/huggingface/diffusers.git`

**Reply by akashlinux10may** [reply]
now everything is sorted as per your instructions, but here its giving new error, ValueError: Could not find the operator torchvision::nms. Please make sure you have already registered the operator and (if registered from C++) loaded it via torch.ops.load_library. is there any way you and me can connect virtually?

**Reply by akashlinux10may** [reply]
see these warning when I am running the 1st cell of your colab
This behaviour is the source of the following dependency conflicts.
torchaudio 2.1.0+cu121 requires torch==2.1.0, but you have torch 2.2.0+cu121 which is incompatible.
torchdata 0.7.0 requires torch==2.1.0, but you have torch 2.2.0+cu121 which is incompatible.
torchtext 0.16.0 requires torch==2.1.0, but you have torch 2.2.0+cu121 which is incompatible.

**Reply by akashlinux10may** [reply]
now I have restarted, cleared my cache, it seems its working fine. training got started.now as the training started, as I have given my own images folder with 5 images, folder name lap, its giving me this error, IsADirectoryError: [Errno 21] Is a directory: '/content/sample_data/lap/.ipynb_checkpoints'

**Reply by asomoza** [reply]
colab makes those hidden directories automatically when you create new ones, just make sure that where you put your images doesn't have anything else on it.

![image](https://github.com/huggingface/diffusers/assets/5442875/8b296511-2d43-44f2-9546-b6a108318ebf)

![image](https://github.com/huggingface/diffusers/assets/5442875/20732ca5-bc09-4e86-bb4a-29a177ee882b)

**Comment by asomoza** [comment]
sorry, I'm actually working right now and try to help when I have time, can't spend that much time on this. 

You're having conflicts between the installations of torch and torchvision, try doing a  `restart session and run all`

![image](https://github.com/huggingface/diffusers/assets/5442875/62706dbb-bb06-44cd-bbf1-a462f82454e1)

if that doesn't work, then something changed in colab between when I did it and now, sadly I can't use colab anymore since I used it for another answer and it doesn't let me run more sessions. When I can connect to it to again I can help more if no else hasn't done it.

**Reply by akashlinux10may** [reply]
I understand, thanks for your help, You already helped me a lot. Much appreciated your time.
